### PR TITLE
add support of RustyHermit's BSD socket layer

### PR DESCRIPTION
**Pull request by stlankes**
RustyHermit is a tier 3 platform and publishes a new kernel interface. The new version supports a common BSD socket layer. By supporting this interface, the implementation of `std` can be harmonized to other operating systems. In `sys_common/mod.rs` we remove only a special case for RustyHermit. All changes are done in the RustyHermit specific directories.

To realize this socket layer, the handling of file descriptors is also harmonized to other operating systems.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1934,8 +1934,16 @@ version = "0.2.6"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "ee512640fe35acbfb4bb779db6f0d80704c2cacfa2e39b601ef3e3f47d1ae4c7"
 dependencies = [
- "compiler_builtins",
  "libc",
+]
+
+[[package]]
+name = "hermit-abi"
+version = "0.3.0"
```
Comment (bjorn3) [inline Cargo.lock:1942]: We will now have all three versions of hermit-abi required at the same time.

Comment (stlankes) [reply Cargo.lock:1942]: Oh, you are right. I will write PRs to upgrade `num_cpus` etc.

**Comment by stlankes** [comment]
The last commit is more a suggestion from my side. I am not sure, where I should store the AsFd and similar traits. I trie to harmonize my implementation to other implementations.

**Comment by Mark-Simulacrum** [comment]
Please mark with `@rustbot ready` when this is ready for review -- looks like maybe some dependency upgrades are still in progress?

**Comment by stlankes** [comment]
Ok, I will do it. However `is-terminal` depends on `io-lifetimes`. To update `io-lifetimes` I need the `fd` traits from the current PR. (https://github.com/hermitcore/io-lifetimes/commit/dd05273e4c13f9d8a757489f8ec865fc972cb703).

**Comment by stlankes** [comment]
@Mark-Simulacrum If I understand it correctly, `atty` is not longer maintained and not all crates moved to `is-terminal`. In addition, `is-terminal` depends indirectly from this PR (see above). How do you handle it?

**Comment by Mark-Simulacrum** [comment]
@bjorn3 is there a reason I'm missing we shouldn't just land this PR, and then follow-up with dropping hermit-abi dependencies once the ecosystem syncs back up?

**Comment by bjorn3** [comment]
That would work.

The changes in files not specific to hermit look correct to me. For the hermit specific files I didn't notice anything obviously wrong, but also didn't review it as closely. Given that RustyHermit is a tier 3 target and the PR author is a maintainer of RustyHermit, I think this is fine.

@bors r+

**Comment by stlankes** [comment]
Do you understand the `homu` error?

File: library/std/src/os/fd/raw.rs [hunk 2]
```diff
@@ -12,11 +15,20 @@ use crate::os::unix::io::OwnedFd;
 #[cfg(target_os = "wasi")]
 use crate::os::wasi::io::OwnedFd;
 use crate::sys_common::{AsInner, IntoInner};
+#[cfg(target_os = "hermit")]
+use hermit_abi::{STDERR_FILENO, STDIN_FILENO, STDOUT_FILENO};
+#[cfg(not(target_os = "hermit"))]
+use libc::{STDERR_FILENO, STDIN_FILENO, STDOUT_FILENO};
```
Comment (mkroening) [inline library/std/src/os/fd/raw.rs:21]: The CI issue happens when documenting std on Windows. `std::os::fd` exists even on Windows with `cfg(doc)`. We might work around like this:

```suggestion
#[cfg(target_os = "hermit")]
use hermit_abi as libc;
```

This, together with changing the `FILENO`s back to `libc::*_FILENO`, should solve the windows docs issue.

**Comment by stlankes** [comment]
@bjorn3 I think that @mkroening's comment is correct. The documentation was broken.

**Pull request merged**